### PR TITLE
Redesign add page to follow masonry layout

### PR DESCRIPTION
**Pull request by nemesis09**
**Jira issues:**
Epic: https://issues.redhat.com/browse/ODC-5013
Story: https://issues.redhat.com/browse/ODC-5561, https://issues.redhat.com/browse/ODC-5417
Bug: https://issues.redhat.com/browse/ODC-5788

**Problem/Description:**
Redesign Add page to support masonry layout

**Solution:**
- Categorize add extensions into groups.
- Add a card for each group in a masonry layout.
- Add data-test-ids wherever required.

**Screens:**
__Add Page redesigned__
add-page-redesign

__Add Page redesigned: responsive__
add-page-redesign-responsive

**Test Coverage:**
Screenshot from 2021-05-10 21-34-05
Screenshot from 2021-05-10 21-35-24
Screenshot from 2021-05-10 21-36-00

**Browser Conformance:**

- [x] Firefox
- [x] Chrome
- [ ] Safari
- [ ] Edge

Note: ~~This PR skips an integration test in kubevirt plugin which checks for Quick starts. It will be reinstated with the PR https://github.com/openshift/console/pull/8950 for Getting Started Section.
https://github.com/openshift/console/blob/1ff04963549964e5f467516622b1e4baf3183597/frontend/packages/kubevirt-plugin/integration-tests-cypress/tests/devconsole/devconsole.spec.ts#L69~~
No change to integration tests. The concerned test has been updated in [PR 8970](https://github.com/openshift/console/pull/8970)

### CODE REVIEW

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/add-actions.ts
Comment (jerolimov) [inline frontend/packages/console-dynamic-plugin-sdk/src/extensions/add-actions.ts:11]: Should we want display an action in more than one group? I would expect that `groupId?: string` is enough?

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/AddCardItem.tsx:22]: You can hide this from screen readers (aria-hidden="true") if this has just the same information (alt text) then the title after this.

```suggestion
      return <img className="odc-add-page__card-item-header-icon" src={icon} alt={label} aria-hidden="true" />;
```

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/AddCardItem.tsx:25]: Same here. I would not expect that an icon add significant infos on a screen reader:

```suggestion
      return <span className={'odc-add-page__card-item-header-icon'} aria-hidden="true">{icon}</span>;
```

File: frontend/packages/dev-console/src/components/add/AddCardSection.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/AddCardSection.tsx:51]: Instead of filtering this action here we should check if we can just remove this action from the `console-extensions.json`, or?

File: frontend/packages/dev-console/src/components/add/AddCardSection.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/AddCardSection.tsx:77]: Please add translations here, for both directions:

```suggestion
            aria-label={showDetails === 'show' ? t('devconsole~Hide add card details') : t('devconsole~Show add card details')}
```

File: frontend/packages/dev-console/src/components/add/AddCardSection.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/AddCardSection.tsx:86]: Please show "Details off" if the switch is disabled. See "User has hidden details" in design doc:

```suggestion
          {showDetails === 'show' ? t('devconsole~Details off') : t('devconsole~Details on')}
```

File: frontend/packages/knative-plugin/console-extensions.json
Comment (jerolimov) [inline frontend/packages/knative-plugin/console-extensions.json:18]: Okay this is the reason to have an array here. We should check this with PM/UX if we want show actions in two different groups.

**Review by jerolimov — CHANGES_REQUESTED** [review]
Really cool for a WIP PR. :+1:

As we discussed on Slack, I started a first review added some small notes:

File: frontend/packages/console-dynamic-plugin-sdk/src/extensions/add-actions.ts
Comment (nemesis09) [reply frontend/packages/console-dynamic-plugin-sdk/src/extensions/add-actions.ts:11]: yes.
`Event sources` is an example. It appears in `Developer Catalog` as well as `Serverless` groups as per the UXD

File: frontend/packages/dev-console/src/components/add/AddCardSection.tsx
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/AddCardSection.tsx:51]: I was thinking on the same lines. Maybe we can have a separate extension for the samples.

**Comment by nemesis09** [comment]
cc @openshift/team-devconsole-ux

File: frontend/packages/dev-console/src/components/EmptyState.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/EmptyState.tsx:54]: Can you please fire this event also on the new action cards.

File: frontend/packages/dev-console/src/components/add/layout/Masonry.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/layout/Masonry.tsx:18]: Also if we developer this together, I now thing we could simplify this and just save the height here.

```suggestion
  const [heights, setHeights] = React.useState<Record<string, number>>({});
  const columns = columnsCount || 1;
  const setHeight = (key: string, height: number) => {
    setHeights((old) => ({ ...old, [key]: height }));
  };
```

File: frontend/packages/dev-console/src/components/add/layout/Masonry.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/layout/Masonry.tsx:29]: With the change above we could simplify this here as well:

```suggestion
      groupedColumns[itemIndex].height += heights[item.key as string] || 0;
```

File: frontend/packages/dev-console/src/components/add/layout/Masonry.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/layout/Masonry.tsx:42]: And here:

```suggestion
      column.height += heights[item.key];
```

**Comment by lwrigh** [comment]
This is looking great to me! Going off of Rachael's design doc it looks like it matches the interactions and behaviors that are documented there.

File: frontend/packages/dev-console/src/components/add/AddCardsLoader.tsx
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/add/AddCardsLoader.tsx:24]: Seems like this component is exactly same as the `EmptyState`. Why do we have two components with same logic?

File: frontend/packages/dev-console/src/components/add/AddCardSection.tsx
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/add/AddCardSection.tsx:34]: Same as before. Why store it as string?

File: frontend/packages/dev-console/src/components/add/AddCardSection.tsx
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/add/AddCardSection.tsx:72]:
```suggestion
              showDetails
```

File: frontend/packages/dev-console/src/components/add/AddCardSection.tsx
Comment (rohitkrai03) [inline frontend/packages/dev-console/src/components/add/AddCardSection.tsx:78]:
```suggestion
              setShowDetails(checked);
```

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/AddCardItem.tsx]: removed redundant key.

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/AddCardItem.tsx:21]: `useUserSettings` didn't exhibit expected behavior while using boolean. Although the value was being stored in storage, but the component didn't re-render resulting in no change in the UI.
cc @jerolimov

Comment (jerolimov) [reply frontend/packages/dev-console/src/components/add/AddCardItem.tsx:21]: The hook was also used at another spot, so it make sense that you extract this into an own hook:

```tsx
export const useShowDetails = (): [boolean, (value: boolean) => void] => {
  const [showDetails, setShowDetails] = useUserSettings(
    'console.addPage.showDetails',
    'show',
    true,
  );
  return [showDetails === 'show', (value: boolean) => setShowDetails(value ? 'show' : 'hide')];
};
```

As we discussed earlier, I now researched the issue that a boolean could not be saved in user settings. It looks like this happen only in the localStorage variant. You can test the "real" user settings (which are stored on the cluster in a ConfigMap) when running the bridge with this parameter. I will create an issue and PR for that so that we can review this independently.

```bash
bin/bridge -user-settings-location configmap
```

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/add/AddCardItem.tsx:22]: How so? What are you thinking here @sahil143 ? This is just a more-complex ternary and we render those without memo all the time.

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/AddCardItem.tsx:45]: Please add a data-test-id here so that we can find this specific add card item.

```suggestion
    <div className="odc-add-card-item" data-test-id={id}>
```

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/AddCardItem.tsx:55]: As the "item" is the complete AddCardItem component I would recommend to call this `link`, so that it's clear which part is "clickable". 

```suggestion
        data-test-id="odc-add-card-link"
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
Checked everything again and added some really small nits. Some of them are opinions and you could check if you want to apply them or not. Please let me know what you think.

**Comment by jerolimov** [comment]
/kind feature

Since the epic acceptance criteria of [ODC-5013](https://issues.redhat.com/browse/ODC-5013) have not changed and the epic has `doc-ack` and `px-ack`, we approve this on behalf of them.
/label px-approved
/label docs-approved

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (sahil143) [reply frontend/packages/dev-console/src/components/add/AddCardItem.tsx:23]: Since this is a method that is being executed during render. we can use the hook use memo which will return the element/component that is needed to render
```
const actionIcon = React.useMemo(() => {
     if (typeof icon === 'string') {
      return (
        <img
          className="odc-add-card-item__header__icon"
          src={icon}
          alt={label}
          aria-hidden="true"
        />
      );
    }
    if (typeof icon !== 'string' && React.isValidElement(icon)) {
      return (
        <span className="odc-add-card-item__header__icon" aria-hidden="true">
          {icon}
        </span>
      );
    }
    return null;
}, [icon, label])
```

Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/AddCardItem.tsx:23]: IMHO, there is not much gain here at the cost of using useMemo, since there is no resource intensive computation.

File: frontend/packages/dev-console/src/components/add/layout/Masonry.tsx
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/layout/Masonry.tsx:13]: using `??` operator would allow falsy values like zero which is not expected for columns here. So I think `||` operator fits the scenario better.

File: frontend/packages/dev-console/src/hooks/hooks.ts
Comment (nemesis09) [reply frontend/packages/dev-console/src/hooks/hooks.ts]: updated the  `useAccessReview` hook

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/AddCardItem.tsx:55]: done

File: frontend/packages/dev-console/src/components/add/AddPageLayout.tsx
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/AddPageLayout.tsx]: done

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/AddCardItem.tsx]: done

File: frontend/packages/dev-console/src/components/add/AddCardItem.tsx
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/AddCardItem.tsx]: done

**Comment by nemesis09** [comment]
cc @openshift/team-devconsole-qe

**Review by jerolimov — APPROVED** [review]
Thanks @nemesis09 for fixing some nits. There are still some open where I will ping you on slack. But its in general fine, but I want wait for @sahil143 feedback here as well.

/approve

Add hold to wait for an `qe-approved` also if Sahil or someone else adds a lgtm!
/hold

**Comment by sahil143** [comment]
Add page not working properly

add-redesign

File: frontend/packages/dev-console/src/components/add/layout/MasonryLayout.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/add/layout/MasonryLayout.tsx:30]: Just found this while debugging the MasonryLayout issue. `columnCount` is not used by the AddCardSection, so you can remove this and optional replace this with an width prop.

```suggestion
type MasonryLayoutProps = {
  columnWidth: number; // not optional, so you need to change AddCardSection as well
  children: React.ReactElement[];
};

export const MasonryLayout: React.FC<MasonryLayoutProps> = ({ columnWidth, children }) => {
  const [width, setWidth] = React.useState<number>(0);
  const columnCount = React.useMemo(() => (width ? Math.floor(width / columnWidth) || 1 : null), [
    columnWidth,
    width,
  ]);

  return (
    <Measure bounds onResize={(contentRect) => setWidth(contentRect.bounds?.width)}>
      {({ measureRef }) => (
        <div className="odc-masonry-container" ref={measureRef}>
          {columnCount ? <Masonry columnsCount={columnCount}>{children}</Masonry> : null}
        </div>
      )}
    </Measure>
  );
};
```

**Comment by nemesis09** [comment]
Updated PR to fix the unexpected behavior and addressed review comments/suggestions.
Please take a look. @sahil143 @jerolimov

**Comment by sanketpathak** [comment]
Tested the PR through cluster-bot. Everything works as expected
Screenshot from 2021-05-19 12-46-37
Screenshot from 2021-05-19 12-46-56

/label qe-approved

File: frontend/packages/dev-console/src/components/add/AddPageLayout.scss
Comment (sahil143) [inline frontend/packages/dev-console/src/components/add/AddPageLayout.scss:20]: nit: missing new line at end of file

**Review by sahil143 — APPROVED** [review]
/lgtm

**Comment by jerolimov** [comment]
Remove the hold as all approvals which are not checked by tide check are available.
/hold cancel

File: frontend/packages/dev-console/src/components/add/AddPageLayout.scss
Comment (nemesis09) [reply frontend/packages/dev-console/src/components/add/AddPageLayout.scss]: There was a trailing space in the new line at the end of the file.
fixed it.

**Pull request merged**
